### PR TITLE
API Restore archive and upload steps

### PR DESCRIPTION
**Pull request by tractorcow**
Fixes https://github.com/silverstripe/cow/issues/54, now that we have recipes implemented. :)

I've also introduced a schema validation utility, based on json-schema.

Restores commands:
  - `cow release:archive`
  - `cow release:upload`
 
Added commands
  - `cow release:wait`
  
Config has been added to installer at https://github.com/silverstripe/silverstripe-installer/commit/4df28ad121987153e0edd45d667b16e04f4a8245

### CODE REVIEW

**Comment by tractorcow** [comment]
@dhensby do you want to review this?

**Comment by dhensby** [comment]
I'm on holiday for another 5 days :/

**Comment by tractorcow** [comment]
Closes https://github.com/silverstripe/silverstripe-framework/issues/7187

**Comment by tractorcow** [comment]
@dhensby do you have time to review this now? :P

File: src/Steps/Release/BuildArchive.php [hunk 1]
```diff
@@ -155,67 +125,31 @@ protected function write($path, $content)
...
+        $version = $archive->getRelease()->getVersion()->getValue();
+        $path = $archive->getTempDir();
+        $this->log($output, "Generating archives for <info>{$name}</info> at <comment>{$path}</comment>");
 
-        // Copy composer.phar
-        $this->log($output, "Getting composer.phar");
-        $this->copy('http://getcomposer.org/composer.phar', "{$path}/composer.phar");
+        // Ensure path is empty, but exists
+        $this->unlink($path);
+        if (!mkdir($path, 0777, true)) {
```
Comment (flamerohr) [inline src/Steps/Release/BuildArchive.php]: does it need to be `0777`? would `0775` or `0755` be sufficient for this temp folder?

File: src/Steps/Release/CreateProject.php [hunk 2]
```diff
@@ -82,6 +82,16 @@ public function run(InputInterface $input, OutputInterface $output)
...
 
+        // Revert composer / recipe-core changes to composer.json
+        $library = $this->getProject();
+        $path = $library->getComposerPath();
+        $repo = $library->getRepository();
+        $status = $repo->run("status", [$path]);
+        if (stripos($status, 'Changes not staged for commit:')) {
+            $this->log($output, "Reverting installer changes to composer.json");
+            $repo->run("checkout", ["--", $path]);
+        }
```
Comment (flamerohr) [inline src/Steps/Release/CreateProject.php]: how come you're reverting? I don't quite follow this step

File: src/Model/Modules/Library.php [hunk 3]
```diff
@@ -1000,18 +1041,10 @@ public function getChangelogPath(Version $version)
...
-            'stabilityVersion' => $version->getStabilityVersion(),
-            'major' => $version->getMajor(),
-            'minor' => $version->getMinor(),
-            'patch' => $version->getPatch(),
-            'version' => $version->getValue(),
-            'versionStable' => $version->getValueStable(),
-        ]);
-        // Collapse duplicate //
-        return str_replace('//', '/', $path);
+        // Substitue version parameters
```
Comment (robbieaverill) [inline src/Model/Modules/Library.php]: Fix for later: typo

File: src/Steps/Release/CreateProject.php (hunk 2 above)
Comment (tractorcow) [reply src/Steps/Release/CreateProject.php]: recipe-module rewrites the root composer.json, but those changes must NOT be committed to source control.

File: src/Steps/Release/BuildArchive.php (hunk 1 above)
Comment (tractorcow) [reply src/Steps/Release/BuildArchive.php]: Fixed.

File: src/Steps/Release/BuildArchive.php [hunk 4]
```diff
@@ -3,72 +3,42 @@
...
-        $path = $this->createProject($output);
-        $this->buildFiles($output, $path);
+        // Get recipes and their versions to wait for
+        $archives = $this->getArchives($output);
+        if (empty($archives)) {
+            $this->log($output, "No recipes configured for archive");
+            return;
+        };
+
+        // Genreate all archives
```
Comment (robbieaverill) [inline src/Steps/Release/BuildArchive.php]: Minor: typo

**Pull request merged**